Explicitly decref in UnpickledPythonCall dtor

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #39555 Release GIL when deleting users and unforked owners
* #39486 Add @rpc.functions.async_execution for rpc.remote
* #39485 Make @rpc.functions.async_execution processing generic
* **#38398 Explicitly decref in UnpickledPythonCall dtor**



Differential Revision: [D21550712](https://our.internmc.facebook.com/intern/diff/D21550712)

Commit:
Minor code cleanup in request_callback_impl

[ghstack-poisoned]

Commit:
Update on "Explicitly decref in UnpickledPythonCall dtor"




[ghstack-poisoned]

xush6528 commented on torch/csrc/distributed/rpc/unpickled_python_call.cpp:
```diff
@@ -14,13 +14,22 @@ UnpickledPythonCall::UnpickledPythonCall(
+  // explicitly setting PyObject* to nullptr to prevent py::object's dtor to
+  // decref on the PyObject again.
+  // See Note [Destructing py::object] in python_ivalue.h
+  py::gil_scoped_acquire acquire;
+  pythonUdf_.dec_ref();
+  pythonUdf_.ptr() = nullptr;
+}
```
By doing this, we can also remove the GIL on line 176 (under MessageType::PYTHON_CALL) and line 287 (under MessageType::PYTHON_REMOTE_CALL).

mrshenli replied:
We still need the GIL there, because `runPythonUdf(...)` returns a new `py::object` and we need the GIL to guard its destruction. Let me add a comment there.

mrshenli commented on torch/csrc/distributed/rpc/python_rpc_handler.cpp:
```diff
@@ -123,13 +123,14 @@ std::shared_ptr<torch::jit::CompilationUnit> PythonRpcHandler::
 
-py::object PythonRpcHandler::runPythonUdf(py::object&& pythonUdf) {
+py::object PythonRpcHandler::runPythonUdf(const py::object& pythonUdf) {
   PROFILE_GIL_SCOPED_ACQUIRE;
```
Yes, making this change just because this PR removed the `movePythonUdf` method and let the UnpickledPythonCall take care of the py::object destruction. So that we no longer need a rvalue arg here. Hence, making this function consistent with other functions in PythonRpcHandler.

Commit:
Update on "Explicitly decref in UnpickledPythonCall dtor"


Differential Revision: [D21550712](https://our.internmc.facebook.com/intern/diff/D21550712)

[ghstack-poisoned]

Commit:
Update on "Explicitly decref in UnpickledPythonCall dtor"


Differential Revision: [D21550712](https://our.internmc.facebook.com/intern/diff/D21550712)

[ghstack-poisoned]

mrshenli committed:
Update on "[WIP] Explicitly decref in UnpickledPythonCall dtor"


Differential Revision: [D21550712](https://our.internmc.facebook.com/intern/diff/D21550712)

[ghstack-poisoned]

mrshenli committed:
Update on "Explicitly decref in UnpickledPythonCall dtor"




Differential Revision: [D21550712](https://our.internmc.facebook.com/intern/diff/D21550712)

[ghstack-poisoned]

mrshenli committed:
Update on "Explicitly decref in UnpickledPythonCall dtor"




Differential Revision: [D21550712](https://our.internmc.facebook.com/intern/diff/D21550712)

[ghstack-poisoned]

mrshenli committed:
Update on "Explicitly decref in UnpickledPythonCall dtor"




Differential Revision: [D21550712](https://our.internmc.facebook.com/intern/diff/D21550712)

[ghstack-poisoned]

mrshenli committed:
Update on "Explicitly decref in UnpickledPythonCall dtor"




Differential Revision: [D21550712](https://our.internmc.facebook.com/intern/diff/D21550712)

[ghstack-poisoned]

mrshenli committed:
Update on "Explicitly decref in UnpickledPythonCall dtor"




Differential Revision: [D21550712](https://our.internmc.facebook.com/intern/diff/D21550712)

[ghstack-poisoned]

mrshenli committed:
Update on "Explicitly decref in UnpickledPythonCall dtor"




Differential Revision: [D21550712](https://our.internmc.facebook.com/intern/diff/D21550712)

[ghstack-poisoned]
